crates/burn-dataset/src/transform/window.rs

```diff
@@ -0,0 +1,277 @@
+    ///
+    /// # Returns
+    ///
+    /// A `WindowsIterator` instance.
+    fn windows(&self, size: usize) -> WindowsIterator<'_, I>;
+}
+
+impl<I, T: Dataset<I>> Windows<I> for T {
+    /// Is empty if the `Dataset` is shorter than `size`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `size` is 0.    
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crate::burn_dataset::{
+    ///    transform::{Windows, WindowsDataset},
+    ///    Dataset, InMemDataset,
+    /// };
+    ///
+    /// let items = [1, 2, 3, 4].to_vec();
+    /// let dataset = InMemDataset::new(items.clone());
+    ///
+    /// for window in dataset.windows(2) {
+    ///  // do sth with window
+    /// }
+    /// ```
+    fn windows(&self, size: usize) -> WindowsIterator<'_, I> {
+        let size = NonZeroUsize::new(size).expect("window size must be non-zero");
+        WindowsIterator::new(self, size)
+    }
+}
+
+/// Overlapping windows iterator.
+pub struct WindowsIterator<'a, I> {
+    /// The size of the windows.
+    pub size: NonZeroUsize,
+    current: usize,
+    dataset: &'a dyn Dataset<I>,
+}
+
+impl<'a, I> WindowsIterator<'a, I> {
+    /// Creates a new `WindowsIterator` instance. The windows overlap.
+    /// Is empty if the input `Dataset` is shorter than `size`.
+    ///
+    /// # Parameters
+    ///
+    /// - `dataset`: The dataset over which windows will be created.
+    /// - `size`: The size of the windows.
+    ///
+    /// # Returns
+    ///
+    /// A `Windows` iterator.
+    pub fn new(dataset: &'a dyn Dataset<I>, size: NonZeroUsize) -> Self {
+        WindowsIterator {
+            current: 0,
+            dataset,
+            size,
+        }
+    }
+}
+
+impl<'a, I> Iterator for WindowsIterator<'a, I> {
+    type Item = Vec<I>;
+
+    fn next(&mut self) -> Option<Vec<I>> {
+        self.current += 1;
+        self.dataset.window(self.current - 1, self.size)
+    }
+}
+
+impl<'a, I> Clone for WindowsIterator<'a, I> {
+    fn clone(&self) -> Self {
+        WindowsIterator {
+            size: self.size,
+            dataset: self.dataset,
+            current: self.current,
+        }
+    }
+}
+
+/// Dataset designed to work with overlapping windows of data.
+pub struct WindowsDataset<D, I> {
+    /// The size of the windows.
+    pub size: NonZeroUsize,
+    dataset: D,
+    input: PhantomData<I>,
+}
+
+impl<D, I> WindowsDataset<D, I>
+where
+    D: Dataset<I>,
+{
+    /// Creates a new `WindowsDataset` instance. The windows overlap.
+    /// Is empty if the input `Dataset` is shorter than `size`.
+    ///
+    /// # Parameters
+    ///
+    /// - `dataset`: The dataset over which windows will be created.
+    /// - `size`: The size of the windows.
+    ///
+    /// # Returns
+    ///
+    /// A `WindowsDataset` instance.
+    pub fn new(dataset: D, size: usize) -> Self
+    where
+        D:,
+    {
+        let size = NonZeroUsize::new(size).expect("window size must be non-zero");
```

Comment:
I considered not panicing here and passing a `NonZeroUsize` instead but I came to the conclusion that not having to declare `NonZeroUsize::new(x)` and `expect("")` or `unwrap()` every single time is worth panicking in the constructor here. 

If you want to avoid panicking in the constructor, I can update the PR accordingly.

Reply:
Nan it's fine.